pkg/threescale/backend/helpers.go

```diff
@@ -0,0 +1,104 @@
+package backend
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/3scale/3scale-go-client/threescale"
+	"github.com/3scale/3scale-go-client/threescale/api"
+)
+
+// blankRequestFromTransactions is used when we need to gather state from the remote backend
+// it parses the transactions for their auth and augments the request to provide nil metrics
+func emptyTransactionFrom(transaction api.Transaction) api.Transaction {
+	return api.Transaction{
+		Metrics:   nil,
+		Params:    transaction.Params,
+		Timestamp: transaction.Timestamp,
+	}
+}
+
+func computeAffectedMetrics(app *Application, request threescale.Request) api.Metrics {
+	return request.Transactions[0].Metrics.AddHierarchyToMetrics(app.metricHierarchy)
+}
+
+// getApplicationFromResponse builds an Application from a response from 3scale client
+func getApplicationFromResponse(resp *threescale.AuthorizeResult) Application {
+	limitCounter := make(LimitCounter)
+
+	for metric, report := range resp.UsageReports {
+		l := &Limit{
+			current: report.CurrentValue,
+			max:     report.MaxValue,
+		}
+		limitCounter[metric] = make(map[api.Period]*Limit)
+		limitCounter[metric][report.PeriodWindow.Period] = l
+	}
+
+	return Application{
+		RemoteState:      limitCounter.deepCopy(),
+		LimitCounter:     limitCounter,
+		UnlimitedCounter: make(map[string]int),
+		metricHierarchy:  resp.AuthorizeExtensions.Hierarchy,
+	}
+}
+
+// getAppIdFromTransaction prioritizes and returns a user key if present, defaulting to app id otherwise
+func getAppIDFromTransaction(transaction api.Transaction) string {
+	appID := transaction.Params.UserKey
+
+	if appID != "" {
+		return appID
+	}
+	return transaction.Params.AppID
+}
```

Comment:
the reason i left that out from the backend client itself is because i thought it better not to force a decision on prioritising where the values should come from